File: Problems/0400_0499/0437_Path_Sum_Three/Project_CS/Path_Sum_Three.cs
using System;

public class Solution
{
    public int PathSum(TreeNode root, int sum)
    {
        if(root == null)
            return 0;
        return includeRoot(root, sum) + PathSum(root.left, sum) + PathSum(root.right, sum);
    }

    public int includeRoot(TreeNode root, int sum)
    {
        if (root == null)
            return 0;
        if (root.val == sum)
            return 1 + includeRoot(root.left, sum - root.val) + includeRoot(root.right, sum - root.val);
        else
            return includeRoot(root.left, sum - root.val) + includeRoot(root.right, sum - root.val);
    }

    public void Main(string args)
    {
        Console.WriteLine("args = " + args );
        string[] flds = args.Replace("\"", "").Replace("[[", "").Replace("]]", "").Trim().Split("],[", StringSplitOptions.None);

        OperateTreeNode ope_t = new OperateTreeNode();
        TreeNode node = ope_t.CreateTreeNode(flds[0]);
        Console.Write("node = \n" + ope_t.TreeToStaircaseString(node));
        Console.WriteLine("node = " + ope_t.Tree2str(node));
        int sum = int.Parse(flds[1]);
        Console.WriteLine("sum = " + sum.ToString());

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        int result = PathSum(node, sum);

        sw.Stop();
        Console.WriteLine("result = " + result.ToString());
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms\n");
    }
}

File: Problems/0400_0499/0437_Path_Sum_Three/Project_Python3/Path_Sum_Three.py
import os
import sys
import time

from TreeNode.Codec import Codec
from TreeNode.TreeNode import TreeNode
from TreeNode.OperateTreeNode import OperateTreeNode

class Solution:
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: int
        """
        if root:
            return self.find_paths(root, sum) + self.pathSum(root.left, sum) + self.pathSum(root.right, sum)
        return 0

    def find_paths(self, root, target):
        if root:
            return int(root.val == target) + self.find_paths(root.left, target - root.val) + self.find_paths(root.right, target - root.val)
        return 0

def main():
    argv = sys.argv
    argc = len(argv)

    if argc < 2:
        print("Usage: python %s <testdata.txt>" %(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("%s not found..." %argv[1])
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        temp = temp.strip()
        if temp == "":
            continue
        print("args = %s" %temp)
        loop_main(temp)
    #    print("Hit Return to continue...")
    #    input()

def loop_main(temp):
    str_args = temp.replace("\"","").replace("[[","").replace("]]","").rstrip()
    flds = str_args.split("],[")

    ope_t = OperateTreeNode()
    root = ope_t.createTreeNode(flds[0])
    print("root = \n{0}".format(ope_t.treeToStaircaseString(root)))
    print("root = {0}".format(ope_t.tree2str(root)))

    sum = int(flds[1].replace("sum = ", ""))
    print("sum = {0:d}".format(sum))

    sl = Solution()
    time0 = time.time()

    result = sl.pathSum(root, sum)

    print("result = %d" %result)

    time1 = time.time()

    print("result = {0:d}".format(result))
    print("Execute time ... : {0:f}[s]\n".format(time1 - time0))

if __name__ == "__main__":
    main()
